devito/finite_differences/operations.py

```diff
@@ -0,0 +1,173 @@
+import sympy
+import numpy as np
+
+import devito
+from devito.symbolics.search import retrieve_functions
+from devito.symbolics.extended_sympy import FrozenExpr
+
+__all__ = ['Mul', 'Add', 'Pow']
+
+
+def fd_parameters(obj):
+    """
+    Process a set of Mul, Add, Pow expression and outputs
+    the space_order, time_order of the whole expression.
+    As the leading order is always the smallest order, if multiple
+    orders are found the lowest is chosen
+    """
+    func = list(retrieve_functions(obj))
+    time_dims = ()
+    sp_dims = ()
+    so = 100
+    to = 100
+    is_time = False
+    stagg = (0, 0, 0, 0)
+    dtype = None
+    # Filter expressions to get space and time Order
+    # The space order is the minimum space order
+    # The time order is the minimum time_order
+    for i in func:
+        if isinstance(i, devito.Function):
+            dtype = i.grid
+            sp_dims += i.space_dimensions
+            so = min(so, i.space_order)
+            stagg = i.staggered
+        if isinstance(i, devito.TimeFunction):
+            dtype = i.grid
+            sp_dims += i.space_dimensions
+            time_dims += (i.time_dim,)
+            so = min(so, i.space_order)
+            is_time = True
+            to = min(to, i.time_order)
+            stagg = i.staggered
+
+    obj.dtype = dtype or np.float32
+    obj.space_order = so
+    obj.time_order = to
+    obj.space_dimensions = tuple(set(sp_dims))
+    obj.staggered = stagg
+    if is_time:
+        obj.time_dim = tuple(set(time_dims))[0]
+    obj.is_TimeFunction = is_time
+
+
+class Pow(sympy.Mul, FrozenExpr):
+    """A customized version of :class:`sympy.Add` representing a sum of
+    symbolic object."""
+    def __new__(cls, *args, **kwargs):
+        return sympy.Pow.__new__(cls, *args, **kwargs)
+
+    def __init__(self, *args, **kwargs):
+        fd_parameters(self)
```

Comment:
exactly -- ``fd_parameters`` should become the constructor of a superclass from which our new ``Pow/Add/Mul/...`` inherit